rest_framework/parsers.py

```diff
@@ -52,17 +53,24 @@ class JSONParser(BaseParser):
     renderer_class = renderers.JSONRenderer
     strict = api_settings.STRICT_JSON
 
+    @staticmethod
+    def _protect_against_surrogate_code_points(decoded_string):
+        m = re.search(r'(?<=[^\\])(?:\\\\)*(\\u[dD][89a-fA-F][0-9a-fA-F]{2})', decoded_string)
```

Comment:
In general this pull request look great - thanks!
Could you walk me through this regex? It might need a comment along with it.

Reply:
Sure! Let me try to describe it to you first and then turn it into an additional in-code comment after your next reply — okay?

So the idea is this:
- `\\u[dD][89a-fA-F][0-9a-fA-F]{2})` matches a four hexdigit unicode code point as specified on https://json.org/. The `u` must be lower (hence no `re.IGNORECASE`) but hexdigits can be lower or upper case. The code point range is surrogates only, else it would be `[0-9a-fA-F]{4}` rather than `[dD][89a-fA-F][0-9a-fA-F]{2}`.
- `(?:\\\\)*` is a non-capturing parenthesis (because we don't need to capture this part) matching an even number of backslashes: 0, 2, 4… That is because JSON could e.g. contain `\\\u1234` to make a backslach and one unicode character or `\\u1234` to make a backslash followed by text "u1234". So even is sane and no surrogates, uneven is surrogates and trouble.
- `(?<=[^\\])` is a positive look-behind for one non-backslash character. Without this, there is no guarantee that `(?:\\\\)*\\` matches an uneven number of slashes, only. So without this, we'd get false positives, rejecting too much.

What do you think?

Reply:
Small update: I have simplified the regex a bit and turned `(?<=[^\\])` into plain `[^\\]` now.

Reply:
@tomchristie I'm rather open about how to address this issue — let's find a way that works well for everyone. Would be great to have your thoughts on https://github.com/encode/django-rest-framework/issues/7026#issuecomment-553491883 if you have a minute.

Reply:
Just tossing this out there - instead of regex matching against the decoded string, would it make sense to re-encode the stream?

Reply:
With `ensure_ascii=False` that would work, yes. This was my first idea too, see section `Workaround` at https://github.com/encode/django-rest-framework/issues/7026#issue-514872212 .